src/boot.c

```diff
@@ -412,31 +427,131 @@ static void read_fsinfo(DOS_FS * fs, struct boot_sector *b, unsigned int lss)
 	fs->free_clusters = le32toh(i.free_clusters);
 }
 
+static int is_fat32(const struct boot_sector *b)
+{
+    return !b->fat_length && b->fat32_length;
+}
+
+static int try_boot_sector(unsigned *boot_sector, unsigned offset,
+                           struct boot_sector *b)
+{
+    unsigned int logical_sector_size;
+    off_t byte_offset = offset * SECTOR_SIZE;
+
+    fs_read(byte_offset, sizeof(*b), b);
+    logical_sector_size = GET_UNALIGNED_W(b->sector_size);
+
+    if (
+	!logical_sector_size ||
+	/* Logical sector size is zero. */
+
+	/* This is the first thing that will fail if the platform needs special
+	 * handling of unaligned multibyte accesses but such handling isn't
+	 * being provided. See GET_UNALIGNED_W() above. */
+	logical_sector_size & (SECTOR_SIZE - 1) ||
+	/* Logical sector size is not a multiple of the physical sector size. */
+
+	!b->cluster_size ||
+	/* Cluster size is zero. */
+
+	!b->fats ||
+	/* Number of FATs is zero. (NTFS is like this.) */
+
+	byte_offset % logical_sector_size
+	/* ...else the boot sector isn't on a logical sector boundary. */
+	)
+    {
+	return 0;
+    }
+
+    *boot_sector = byte_offset / logical_sector_size;
+    return 1;
+}
+
+static int has_signature(const void *sector)
+{
+    const uint8_t *sector_u8 = sector;
+    return sector_u8[510] == 0x55 && sector_u8[511] == 0xaa;
+}
+
+static int try_backup_boot_sector(unsigned *boot_sector, unsigned offset,
+                                  struct boot_sector *b)
+{
+    /* Backup boot sectors are for FAT32 only. Also, check backup boot sectors
+     * for a signature. */
+    return
+	try_boot_sector(boot_sector, offset, b) &&
+	is_fat32(b) &&
+	has_signature(b);
+}
+
 void read_boot(DOS_FS * fs)
 {
     struct boot_sector b;
+    unsigned boot_sector;
     unsigned total_sectors;
     unsigned int logical_sector_size, sectors;
     long long fat_length;
     unsigned total_fat_entries;
     off_t data_size;
     long long position;
 
-    fs_read(0, sizeof(b), &b);
+    if (!try_boot_sector(&boot_sector, 0, &b)) {
+	printf("Boot sector is unusable.\n");
+
+	/* Backup boot sector is at sector 6 by default. Check that first, for
+	 * valid sector sizes of 512 to 4096 bytes.
+	 *
+	 * Might not be a bad idea to only check this for disks large enough to
+	 * accommodate FAT32. */
+	if (!try_backup_boot_sector(&boot_sector, 6, &b) &&
+	    !try_backup_boot_sector(&boot_sector, 12, &b) &&
+	    !try_backup_boot_sector(&boot_sector, 24, &b) &&
+	    !try_backup_boot_sector(&boot_sector, 48, &b)) {
+	    unsigned sector = 1;
+	    for (;;) {
+		if (sector == 32)
+		    die("Backup boot sector not found.");
+		if (sector != 6 &&
+		    sector != 12 &&
+		    sector != 24 &&
+		    try_backup_boot_sector(&boot_sector, sector, &b)) {
+		    break;
+		}
+		sector++;
+	    }
+	}
+    }
+
     logical_sector_size = GET_UNALIGNED_W(b.sector_size);
```

Comment:
That check is in `try_boot_sector`, the `byte_offset % logical_sector_size` bit.

The `offset` parameter in `try_(backup_)boot_sector` is a multiple of 512 (`SECTOR_SIZE`), because any other sector size will be a multiple of that. For example, with a logical sector size of 4 KB, it might find a backup boot sector at 48 * 512, then right after it'll do a `GET_UNALIGNED_W(b.sector_size)`, and `boot_sector` will be set to 48 * 512 / 4096 == 6.

Reply:
Ok.